[INTEL MKL] Added support for Quantized-Conv2D and Pad fusion.

bhavani-subramanian opened the pull request.

bhavani-subramanian committed:
Enabled quantized pad fusion.

shahzadlone commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -467,19 +467,18 @@ class MklConvOp : public OpKernel {
                                         filter.shape().DebugString()));
 
     for (int i = 0; i < 3; i++) {
```
I know you didn't do this but could it change this ```i++``` to ```++i``` while at it, please..

bhavani-subramanian replied:
Done. Thanks!

penpornk reviewed (changes requested):
Thank you very much for the PR!

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -2074,7 +2089,7 @@ REGISTER_KERNEL_BUILDER(
                               .Label(mkl_op_registry::kMklOpLabel),        \
                           MklConvOp<CPUDevice, float, float, float, float, \
-                                    float, int64, false, true, false>);    \
+                                    float, int32, false, true, false>);    \
```
Since `padding_list` is of type `vector<int32>` in the layout pass, `Tpaddings` needs to be `int32` as well - otherwise the compiler complains about not being able to convert `int*` to `long long int*`  in line 1167:
```
1166 if (quantized_pad_enabled) {
1167  paddings = padding_list_.data();
1168} else {
```

penpornk replied:
But then why do you need the type constraint 3 lines above to be `int64`? 
https://github.com/tensorflow/tensorflow/blob/0a507dc55b51e4f5651b57473ff5e2b3ea50d95e/tensorflow/core/kernels/mkl_conv_ops.cc#L2090

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -2116,7 +2131,7 @@ TF_CALL_float(REGISTER_MKL_CPU_2D_DEPTHWISE);
           .TypeConstraint<int64>("Tpaddings")                       \
           .Label(mkl_op_registry::kMklOpLabel),                     \
-      MklFusedConvOp<CPUDevice, T, T, T, T, T, int64, true>);       \
+      MklFusedConvOp<CPUDevice, T, T, T, T, T, int32, true>);       \
```
Same comment as above.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc:
```diff
@@ -0,0 +1,454 @@
+
+  TF_ASSERT_OK(RunOpKernel());
+
+  // Output -> int32
```
Good catch! It should be `qint32`.

bhavani-subramanian committed:
Addressed review comments.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc:
```diff
@@ -0,0 +1,454 @@
+  const int stride = 1;
+  ConfigureQuantizedConv2D(stride);
+
+  const int depth = 1;
```
Done.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -938,9 +935,18 @@ class MklConvOp : public OpKernel {
+      for (auto const& padding_val : padding_list_) {
+        if (!padding_val) continue;
+        quantized_pad_enabled = true;
+        break;
```
Agree. Changed it per your suggestion.

bhavani-subramanian replied:
@penpornk Just for my learning purposes, can you help me understand how having a `continue` results in more jumps? Any pointers to material would also help. Thanks!

penpornk replied:
Sorry about the confusion! The codes should have the same number of jump instructions. I was just thinking if one might need to jump less based on what values padding_val usually have. I should have added "probably" because I didn't try to count properly and was just speculating what padding_list_ would be like.
```
  compare_condition
  jump_to_"false_label"_if_comparison_is_false
  do_something_if_true
false_label:
  commands_outside_if
```

bhavani-subramanian replied:
No worries. Thanks for the explanation!

bhavani-subramanian commented on tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc:
```diff
@@ -0,0 +1,454 @@
+
+  // Output -> float
+  const int expected_width = image_width;
+  const int expected_height = image_height * filter_count;
```
Done.

bhavani-subramanian commented:
@penpornk Thanks for quickly reviewing the PR! I have addressed your review comments. Let me know if you need anything else.

penpornk reviewed (commented):
Thank you for your quick changes!

bhavani-subramanian committed:
Changed type of 'padding_list_' to 'Tpadding'. Added additional registrations for
QuantizedConv2D + bias + sum (and signed sum) + relu & requantize fusion for when
Tbias is a float.

bhavani-subramanian commented:
@penpornk I have changed `padding_list_` to be of type `Tpadding` instead of `int32`.

Also, I have added an additional registration for when `Tbias` is a `float` since we support both `qint32` and `float` types for Tbias for `Quantized-Conv2D + bias + sum + relu + requantize` fusion.

penpornk reviewed (approved):
Looks good. Thank you again for the PR! :)

bhavani-subramanian commented:
@pragyaak The unit test failures don't seem to be related to the changes made in this PR. Please let me know if you think otherwise.

penpornk reviewed (changes requested):
@bhavani-subramanian The `Ubuntu Python2` and `Ubuntu Python3` failures are actually related to this PR. The API golden files need to be modified because we added an extra attribute `padding_list` to the ops. See the `Ubuntu Python 2` test log [here](https://source.cloud.google.com/results/invocations/be1124ab-1bf9-4542-b3c8-f61de7af5a4b/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log).

Could you please run the following commands and add the modified golden files (`tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt`, and `tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt`)to the PR?
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
After doing this, your local `bazel test tensorflow/tools/api/tests:api_compatibility_test` should pass. 
Thank you!

bhavani-subramanian committed:
Updated the API golden files since 'padding_list' attribute
was added to quantized Conv2D ops.

bhavani-subramanian commented:
Thanks for the explanation, @penpornk! I tried modifying the API golden files per your suggestion, but only `tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt` ended up getting updated, and `tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt` remains unmodified. However, I do see that `bazel test tensorflow/tools/api/tests:api_compatibility_test` unit test passes locally even with just this change.

penpornk commented:
@bhavani-subramanian We pulled a PR in from last week with just changes to `v2`, but it failed the internal test for `v1` and we had to add changes to `v1` anyway. So I think we should add it now. I'm not sure how to get the OSS test to modify `v1`, though. (Apart from modifying it manually from the diff.) 

I'm asking @annarev: 
Do you know how to make the api_compatibility_test also modify the `v1` file? Thank you!

annarev commented:
I submitted a fix for that issue last week. @bhavani-subramanian can you sync to head and check if api_compatibility_test updates both v1 and v2 files now?

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/quantized-pad-fusion-latest

bhavani-subramanian committed:
Added v1 golden file for 'api_compatibility_test'

bhavani-subramanian commented:
Thanks for the quick response, @annarev. I was able to add the `v1` file after merging my branch with the master.

@penpornk Please let me know if anything else is needed from my end. Thanks!

penpornk reviewed (approved):
@bhavani-subramanian I think it looks good now. Thank you so much for the changes! (And thank you, @annarev!)

Re-running the tests and waiting for API review then.

bhavani-subramanian commented:
@penpornk The two failures in Ubuntu contrib pass when run locally. Not sure how to reproduce the error at my end.

penpornk commented:
@bhavani-subramanian I'll rerun the tests. I can't see the log of `Ubuntu Python 2`. The rest looks unrelated.

bhavani-subramanian commented:
@penpornk Sorry to bug you again. The unit test that fails in `MacOS Python2 and CC` (`//tensorflow/python/saved_model:load_test`) passes when run locally. From the error log, it looks like the test failed because it was unable to create a new thread either because the system either ran out of resources (or) due to hitting a system limit on the number of threads per process ([Reference](https://stackoverflow.com/questions/26691340/multi-threading-in-c-throws-thread-constructor-failed-resource-temporarily-un)).

It doesn't seem like this failure is related to my PR.

penpornk commented:
@bhavani-subramanian Ah. Don't worry about them! All three failing tests right now are existing failures (`MacOS Python2 and CC`, `Windows Bazel`, and `Windows Bazel GPU`) I mainly reran the tests because `Ubuntu Python 2` and `Ubuntu Contrib` shouldn't be failing. They are passing now so I think we are fine.

I haven't tagged `ready to pull` yet because this still needs API review (because we changed the golden files).

Srini511 commented:
@pragyaak gentle reminder regarding the API review for this PR.

penpornk commented:
Apologies for the delay, but this PR finally got through the API review and is getting merged internally. But we got outdated files, specifically,
```
//depot/google3/third_party/tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt
//depot/google3/third_party/tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt
//depot/google3/third_party/tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt
```
Could you please help resolve these? Thank you!

bhavani-subramanian commented:
@penpornk I merged my branch with the master and ran the commands below. But the above files are still not getting updated:
```
 $ bazel build tensorflow/tools/api/tests:api_compatibility_test
 $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
Are there other commands I should be running instead?

penpornk commented:
@bhavani-subramanian Do you have the changes from the latest commit on these two files?: https://github.com/tensorflow/tensorflow/commit/4294edea3e80c746eac400262628f852b66d02e5

penpornk commented:
@bhavani-subramanian Never mind! It's getting merged now. :)

bhavani-subramanian commented:
Thanks @penpornk! I am somehow not able to generate the golden files despite having the `4294ede` commit. Anyway, glad that its getting merged now :)
